task/move.go

```diff
@@ -86,27 +87,79 @@ func Move(fetch Fetcher, put Putter) {
 	}
 }
 
+func (o *oldStore) fetchB() (thing Thing, ok bool, err error) {
+	// simulate the delay
+	time.Sleep(2 * time.Second)
+	fmt.Println("fetchB: fetching...", o.bookNo)
+
+	if o.bookNo == 5 {
+		return thing, ok, errors.New("Fetch encountered network issues")
+	}
+
+	if o.bookNo > len(o.bookInventory)-1 {
+
+		return thing, ok, err
+	}
+
+	time.Sleep(1 * time.Second)
+	thing, ok = o.bookInventory[o.bookNo], true
+	o.bookNo++
+	return
+}
+
+func (n *newStore) putB(thing Thing) error {
+	time.Sleep(5 * time.Second)
+
+	n.inventory = append(n.inventory, thing)
+
+	if len(n.inventory) == 4 {
+		return errors.New("store is filled up")
+	}
+
+	return nil
+}
+
 // MaybeMove is exactly the same as Move except that it may return an error
 // because fetch() and put() may return errors. If no errors occur then
 // MaybeMove returns under the same conditions as Move(). If an error occurs
 // then MaybeMove returns earlier even if there are more Things to fetch().
 func MaybeMove(fetch MaybeFetcher, put MaybePutter) error {
-	// [TODO]
+	ch := make(chan Thing)
+	errCh := make(chan error, 1)
+	done := make(chan struct{}, 1)
```

Comment:
Usually we call this channel `quit` because it is telling the goroutine to quit, and we reserve the name `done` for a channel that is used to signal that something is, well, done :)

```Go
quit := make(chan struct{}) // what you're calling quit
done := make(chan struct{})

go func(){
  defer close(done)
  select {
    case <-quit:
      return
    // other cases
  }
}()

close(quit) // makes goroutine stop

// Waiting to receive on a signalling channel is a very common idiom
// that you'll see. We're never going to send on it, so the only time we'd
// ever receive on a <-done is when the channel is closed.
<-done
```

Reply:
There's also no need to buffer it and then send as a signal. Rather just close the channel like I did in my example.